### ISSUE TITLE
UMap in YOLO

### ISSUE THREAD

**Issue by Abolfazl-kr** [ultralytics/ultralytics#17443]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone.

I've trained a Yolo classification with my custom dataset. now I want to plot the UMap to show how the trained model classify my data.

please tell me how I could do this, a sample of UMap is here:

![This-figure-shows-the-UMAP-manifolds-of-features-extracted-from-a-trained-YOLO-model-on_W640](https://github.com/user-attachments/assets/4e91b43a-9e0a-4949-b0f2-f10b62d80005)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Abolfazl-kr, thank you for reaching out to Ultralytics and showing interest in enhancing your YOLO model's analysis with UMap visualizations 🚀! 

We recommend you check out our [Docs](https://docs.ultralytics.com) where you'll find various resources and examples for different tasks, including the use of Python and CLI to manipulate and analyze models. For visualizations, particularly like UMap, it's beneficial to explore the sections on model evaluation and inferential statistics.

In case this involves a 🐛 Bug or if you encounter any issues while attempting your UMap integration, kindly provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) — this would greatly assist us in understanding and resolving the issue.

Additionally, if you need real-time support or wish to engage with other Ultralytics enthusiasts and experts, feel free to join our [Discord](https://ultralytics.com/discord) community 🎧. For more structured discussions, consider visiting [Discourse](https://community.ultralytics.com), or engage with our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're using the latest features and fixes by upgrading to the newest `ultralytics` package and related [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) setup, with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

The latest YOLO deployments can be tested in various environments that come pre-configured and optimized for intense computational tasks:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Explore more in our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. Details are available in our [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge reflects the current status of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests. They run every 24 hours and with every commit to ensure all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) function correctly across platforms.

Please note that this is an automated response, and an Ultralytics engineer will review your issue shortly 🛠️. Thank you for your patience and contribution!

**Comment by glenn-jocher** [comment]
@Abolfazl-kr to visualize UMAP embeddings from your trained YOLO model, you can extract features from the model's penultimate layer and then use a library like UMAP-learn in Python to generate the plot. If you need further assistance, please refer to the UMAP documentation for detailed guidance.

**Comment by Abolfazl-kr** [comment]
@glenn-jocher 

Thanks a lot for your response.

How can i get access to the model's penultimate layer? Is it a 2D array like feature maps or a vector?

**Comment by glenn-jocher** [comment]
@Abolfazl-kr you can access the model's penultimate layer by modifying the forward pass to return the desired layer's output, which is typically a vector representing features. If you need specific code examples, feel free to ask!

**Comment by Abolfazl-kr** [comment]
@glenn-jocher 

Yes, I would appreciate it if you could send me the example code to access the model's penultimate layer.

**Comment by glenn-jocher** [comment]
Certainly! You can modify the model's forward method to return the output of the penultimate layer. Here's a minimal example:

```python
import torch
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Modify the forward pass
def forward_hook(module, input, output):
    global penultimate_output
    penultimate_output = output

# Register the hook
layer = model.model[-2]  # Access the penultimate layer
hook = layer.register_forward_hook(forward_hook)

# Run inference
_ = model('path/to/image.jpg')

# Access the penultimate layer output
print(penultimate_output)

# Remove the hook
hook.remove()
```

This code snippet demonstrates how to capture the output of the penultimate layer during inference. Adjust the layer index as needed based on your model architecture.

**Comment by Abolfazl-kr** [comment]
@glenn-jocher 

Thanks a lot.

Actually, i'm trying to plot UMap in a classification problem, but when I run the code I get this error.

layer = model.model[-2]
TypeError: 'ClassificationModel' object is not subscriptable

**Comment by glenn-jocher** [comment]
@Abolfazl-kr it seems like the model object doesn't support direct indexing. You can access the layers by iterating over `model.model` or using `model.model.children()`. Adjust the code to access the desired layer accordingly.
